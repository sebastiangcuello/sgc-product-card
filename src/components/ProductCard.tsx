import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct'
import { InitialValues, Product, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    //children?: ReactElement | ReactElement[];
    children: ( args?: ProductCardHandlers ) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {
  
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ 
        onChange,
        product,
        value,
        initialValues
    })

    return (
        // es un componente de react que permite pasar datos a los componentes hijos
        <Provider value={{
            counter,
            increaseBy,
            maxCount,
            product
        }}> 
            <div 
                className={ `${ styles.productCard } ${ className } ` }
                style= { style }
            >
                { children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset
                }) }
              </div>
        </Provider>

    )
}

// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;