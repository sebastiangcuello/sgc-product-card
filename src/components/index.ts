import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductCardHOCProps } from "../interfaces/interfaces";

import { ProductButtons } from "./ProductButtons";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

//Con Object.assign podemos asignarle propiedades a un objeto
//En este caso le asignamos las propiedades Image, Title y Buttons
export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Image: ProductImage,
    Title: ProductTitle,
    Buttons: ProductButtons,
});

export default ProductCard;