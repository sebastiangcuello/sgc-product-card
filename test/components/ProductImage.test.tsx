
import React from "react";
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from "../../src/components";
import { product2 } from "../data/product";

describe('ProductImage', () => { 
    
    test('should render the image', () => {

        const wrapper = renderer.create(
            <ProductImage img="./coffee-mug.png" className="class-prueba" />
        );

        expect( wrapper.toJSON() ).toMatchSnapshot();
    });

    test('should show the component with image', () => {

        const wrapper = renderer.create(
            <ProductCard product={ product2 }>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect( wrapper.toJSON() ).toMatchSnapshot();

    })
 })