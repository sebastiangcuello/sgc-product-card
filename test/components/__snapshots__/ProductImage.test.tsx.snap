// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should render the image 1`] = `
<img
  alt="Product"
  className="productImg class-prueba "
  src="./coffee-mug.png"
/>
`;

exports[`ProductImage should show the component with image 1`] = `
<div
  className="productCard undefined "
>
  <img
    alt="Product"
    className="productImg undefined "
    src="./coffee-mug.png"
  />
</div>
`;
